framework/data/BaseDataProvider.php

```diff
@@ -165,7 +168,14 @@ public function getTotalCount()
     {
         if ($this->getPagination() === false) {
             return $this->getCount();
-        } elseif ($this->_totalCount === null) {
+        }
+
+        return $this->getTotalCountWithPagination();
+    }
+
+    private function getTotalCountWithPagination()
```

Comment:
Extracting this method doesn't add much clarity. Especially the method name.

Also there's not much to reuse. I'd revert it and, instead, add two lines to `getPagination()` directly.

Reply:
Yes, we can use `$this->_pagination->totalCount = $this->prepareTotalCount();` instead of `$this->_pagination->totalCount = $this->getTotalCountWithPagination();`. But we will calc totalCount twice, one is calculating Pagination.totalCount, the other is  calculating DataProvider.totalCount.

Reply:
@samdark I made it, as you said.